### PR TITLE
MTT-4370: UTP 2.0 migration guide

### PR DESCRIPTION
**Pull request by simon-lemay-unity**
**Select the type of change:**

- [ ] Small Changes - Typos, formatting, slight revisions
- [ ] New Content - New features, sections, pages, tutorials
- [ ] Site and Tools - Updates, maintenance, and new packages for the site and Docusaurus

**Purpose of the Pull Request:**

This PR introduces the migration guide when updating from UTP 1.X to 2.0. The intent is to leave it as a draft for now since we might have more breaking changes to add in there as we move closer to the release.

Before merging, I'll also update the PR so that most of the APIs link to their API page. I didn't do so right away to make the markdown file easier on the eyes while we review it.

**Issue Number:**  MTT-4370

### CODE REVIEW

File: transport_versioned_docs/version-2.0.0/migration.md [hunk 1]
```diff
@@ -1,9 +1,46 @@
...
 
+Note that for most use cases, no action should be required when updating to 2.0. The core APIs like `NetworkDriver` remain the same, with the biggest changes limited to specialized scenarios (e.g. custom network interfaces).
+
+## Protocol incompatibility
+
+The custom communication protocol used by UTP to implement connections over UDP has changed in a backward-incompatible manner. This means that clients running UTP 2.0 or later can't connect to servers running 1.X, and vice versa. Attempting to establish such connections will result in connection failure after the usual timeout.
+
+It is thus recommended to ensure that clients and servers are updated at the same time, or to disallow older clients from connecting once the server is updated. Another strategy is to offer different endpoints for UTP 1.X and 2.0 servers, which could smooth the transition while older clients are updated.
+
+The reasons for this breaking change are to improve bandwidth efficiency, simplify the protocol, and (ironically) improve forward-compatibility of the protocol.
```
Comment (johann-r-unity) [inline transport_versioned_docs/version-2.0.0/migration.md:17]: The "(Ironically)" piece contrast with the doc. Could we use something like : 
and lays the foundations for better forward compatibility of the protocol.

File: transport_versioned_docs/version-2.0.0/migration.md [hunk 2]
```diff
@@ -1,9 +1,46 @@
...
+
+  * Instead of registering a custom pipeline stage with `NetworkPipelineStageCollection.RegisterPipelineStage`, use `NetworkDriver.RegisterPipelineStage`. Note that this must be performed on every instance of `NetworkDriver` that will make use of the custom pipeline stage.
+  * Instead of retrieving the ID of a pipeline stage with `NetworkPipelineStageCollection.GetStageId`, use the static `NetworkPipelineStageId.Get` method.
+
+The reason for these breaking changes is to remove Burst-incompatible APIs, allowing more of UTP to be used from Burst-compiled code. For more details on how to create custom pipeline stage, please refer to [this section of the documentation](custom-pipeline.md). For more details on pipelines in general, see [this section instead](pipelines-usage.md).
+
+## Other breaking changes
+
+  * After calling `NetworkDriver.Disconnect`, notifying the remote peer of the disconnection now requires completing a `NetworkDriver.ScheduleUpdate` job. In 1.X, `NetworkDriver.ScheduleFlushSend` used to be sufficient for this purpose but it's not the case anymore.
```
Comment (johann-r-unity) [inline transport_versioned_docs/version-2.0.0/migration.md:42]: All the document is factual without explanation, I think this is the correct approach. 
However, for this, we are adding an extra step for the user. The natural reaction will be to wonder why. 
Would an explanation make sense here ?

**Review by johann-r-unity — APPROVED** [review]

File: transport_versioned_docs/version-2.0.0/migration.md [hunk 3]
```diff
@@ -1,9 +1,46 @@
...
+
+  * There is no concept of `NetworkInterfaceEndPoint` anymore. It has been completely replaced with the more general `NetworkEndpoint`. Consequently, there is no need to implement conversion logic between the two anymore (so `CreateInterfaceEndPoint` and `GetGenericEndPoint` were removed from `INetworkInterface`).
+  * There is no need to provide a `NetworkSendInterface` through `CreateSendInterface` anymore. Send operations are now handled entirely by `ScheduleSend`, which gets passed a `PacketsQueue` containing the packets to be sent.
+  * The `ScheduleReceive` method doesn't use `NetworkPacketReceiver` (which is now obsolete) to propagate received packets to the rest of UTP. Instead, implementations of `ScheduleReceive` are expected to fill the `PacketsQueue` that is now passed in with the received packets.
+  * Implementations of `INetworkInterface` are now expected to be fully unmanaged. However, a managed implementation can be wrapped into an unmanaged one with the new `WrapToUnmanaged` extension method.
```
Comment (wackoisgod) [inline transport_versioned_docs/version-2.0.0/migration.md:26]: So the term unmanaged I think will leave folks confused. It almost sounds like the interface must be written in native code which isn't really what you are saying. What you are saying is that it must be fully blittable unless you are using the wrapper.

**Comment by wackoisgod** [comment]
Is 2.0 going to depend on 2.0 of collections? I noticed that in that version of collections DataStream and NetworkCompression are now part of collections.

**Comment by johann-r-unity** [comment]
@wackoisgod , yes.

**Comment by wackoisgod** [comment]
> @wackoisgod , yes. 

If that's the case then it may be worth it to call out those changes. Also if I recall collections 2.0 requires Unity 2022.2 according to the preview release. Does that mean 2.0 will require that?

**Comment by johann-r-unity** [comment]
You are correct. It will be available with the Editor starting 2022.2. 
This migration guide is gonna be part of a larger effort to update the docs, but I agree that mentioning it on this page as well makes sense.

**Comment by wackoisgod** [comment]
> You are correct. It will be available with the Editor starting 2022.2. 
> 
> This migration guide is gonna be part of a larger effort to update the docs, but I agree that mentioning it on this page as well makes sense. 

So does this mean you will continue to support 1.x and make fixes and updates for that version? My worry is that 2.0 will be out while two LTS versions are out under 1.x since it doesn't seem things are backwards compatible.

**Comment by simon-lemay-unity** [comment]
> So does this mean you will continue to support 1.x and make fixes and updates for that version? My worry is that 2.0 will be out while two LTS versions are out under 1.x since it doesn't seem things are backwards compatible.

Yes. As long as there are supported LTS versions that only support UTP 1.X, we'll keep supporting it. Some new features may only make it in 2.0, however.

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=752) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=752) before we can accept your contribution.<br/>**0** out of **4** committers have signed the CLA.<br/><br/>:x: Grave18<br/>:x: Ianisthebest<br/>:x: dependabot[bot]<br/>:x: Briancoughlin<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=752) it.</sub>

File: transport_versioned_docs/version-2.0.0/migration.md (hunk 2 above)
Comment (armstrongl) [reply transport_versioned_docs/version-2.0.0/migration.md]: @johann-r-unity - Forgive me for lacking context here; I'm not sure I understand entirely. Are you saying we should add an explanation around why we introduced these additional breaking changes (similar to the explanation at the end of the previous paragraphs)?

For example:
> These breaking changes simplify and increase the flexibility of the interface. Please refer to this documentation section for more details on creating custom network interfaces.

Comment (johann-r-unity) [reply transport_versioned_docs/version-2.0.0/migration.md]: Hi Larah, 
My initial comment is outdated. Simon did the required modification already.

**Review by johann-r-unity — APPROVED** [review]

**Review by armstrongl — APPROVED** [review]

**Pull request merged**